contracts/gateway/EIP20Gateway.sol

```diff
@@ -472,6 +472,14 @@ contract EIP20Gateway is GatewayBase {
 
         // Get the message object
         MessageBus.Message storage message = messages[_messageHash];
+        // Return revert penalty to staker if message is already progressed
+        // and can't be reverted anymore.
+        tryReturnPenaltyToStaker(
```

Comment:
As a general practice, the token transfer should be done after the state change.  This should be done after progressStakeInternal,  I think.